components/support/migration/src/main/java/mozilla/components/support/migration/FennecMigration.kt

```diff
@@ -0,0 +1,216 @@
+            val migrationVersion = tuple.getInt("version")
+            val migrationSuccess = tuple.getBoolean("success")
+            val migration = when (migrationName) {
+                "HISTORY" -> Migration.HISTORY
+                "BOOKMARKS" -> Migration.BOOKMARKS
+                "OPEN_TABS" -> Migration.OPEN_TABS
+                else -> throw IllegalStateException("Unrecognized migration type")
+            }
+            result[migration] = MigrationRun(version = migrationVersion, success = migrationSuccess)
+        }
+        return result
+    }
+}
+
+// Used to ensure migration runs do not overlap.
+private val migrationLock = Object()
+
+/**
+ * Fennec data migration scaffolding.
+ */
+class FennecMigration(
+    private val context: Context,
+    private val historyStorage: PlacesHistoryStorage,
+    private val bookmarksStorage: PlacesBookmarksStorage,
+    // Single-thread executor to ensure we don't accidentally parallelize migrations.
+    private val coroutineContext: CoroutineContext = Executors.newSingleThreadExecutor().asCoroutineDispatcher() + SupervisorJob()
+) {
+    private val logger = Logger("FennecMigration")
+
+    private val profile = FennecProfile.findDefault(context)
+    private val dbPath = profile?.let { "${it.path}/browser.db" }
+
+    fun migrateAsync(): Deferred<MigrationsResult> = synchronized(migrationLock) {
+        val migrationRecord = MigrationHistoryStore(context)
+        val migrationHistory = migrationRecord.getCached()
+
+        // Either didn't run before, or ran with an older version than current migration's version.
+        val migrationsToRun = migrations.filter {
+            val pastVersion = migrationHistory?.get(it.first)?.version
+            if (pastVersion == null) {
+                true
+            } else {
+                it.second > pastVersion
+            }
+        }
+
+        // Short-circuit if there's nothing to do.
+        if (migrationsToRun.isEmpty()) {
+            logger.debug("No migrations to run.")
+            val result = CompletableDeferred<MigrationsResult>()
+            result.complete(emptyMap())
+            return result
+        }
+
+        return runMigrationsAsync(migrationsToRun)
+    }
+
+    private fun runMigrationsAsync(
+        migrations: List<Pair<Migration, Int>>
+    ): Deferred<MigrationsResult> = CoroutineScope(coroutineContext).async { synchronized(migrationLock) {
+        val results = mutableMapOf<Migration, Pair<Int, MigrationResult>>()
+
+        migrations.forEach { versionedMigration ->
+            logger.debug("Migrating $versionedMigration")
+
+            val migrationResult = when (versionedMigration.first) {
+                Migration.HISTORY -> migrateHistory()
+                Migration.BOOKMARKS -> migrateBookmarks()
+                Migration.OPEN_TABS -> migrateOpenTabs()
+            }
+
+            results[versionedMigration.first] = versionedMigration.second to migrationResult
```

Comment:
nit: The usage of `first` and `second` and a map with a pair makes it hard to follow this code.

Reply:
Ah yeah, pairs kind of suck :(